File: app/lib/workspace.js
/**
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership.
 *
 * Camunda licenses this file to you under the MIT; you may not use this file
 * except in compliance with the MIT License.
 */

const { readFile, readFolderStats } = require('./file-system');

const renderer = require('./util/renderer');

const log = require('./log')('app:workspace');

const { forEach, isString } = require('min-dash');


class Workspace {
  constructor(config) {
    renderer.on('workspace:restore', (defaultConfig, done) => {
      const workspace = config.get('workspace', null);

      if (!workspace) {
        return done(null, defaultConfig);
      }

      let files = [];

      // ensure backwards compatibility
      forEach((workspace.files || workspace.tabs), file => {
        const { path } = file;

        try {
          log.info(`restoring ${ path }`);

          files = [
            ...files,
            readFile(path)
          ];
        } catch (error) {
          log.error(`failed to restore ${ path }`, error);
        }
      });

      workspace.files = files;

      workspace.endpoints = workspace.endpoints || [];

      if (workspace.project) {
        if (isString(workspace.project)) {
          workspace.project = readFile(workspace.project);
        }

        workspace.project.folders.forEach((folderPath, index) => {
          const folder = readFolderStats(folderPath);

          workspace.project.folders[ index ] = folder;
        });
      }

      done(null, {
        ...workspace
      });
    });

    renderer.on('workspace:save', (workspace, done) => {
      log.info('saving');

      if (workspace.project) {
        if (workspace.project.path) {
          workspace.project = workspace.project.path;
        } else {
          workspace.project.folders = workspace.project.folders.map(folder => folder.path);
        }
      }

      config.set('workspace', workspace);

      done(null);
    });
  }
}

module.exports = Workspace;

File: src-tauri/crates/modeler-backend/src/workspace.rs
// Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
// under one or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information regarding copyright
// ownership.
//
// Camunda licenses this file to you under the MIT; you may not use this file
// except in compliance with the MIT License.

//! Workspace restore/save, a faithful port of `app/lib/workspace.js`.
//!
//! The workspace is stored under the `workspace` config key. On restore, each
//! saved file is re-read from disk so the renderer gets fresh contents;
//! unreadable files are skipped individually (NOT fatal), and a saved workspace
//! with zero readable files still returns the workspace object (with an empty
//! `files` list) rather than the default.

use serde_json::Value;

use crate::config::Config;
use crate::error::IpcError;
use crate::file_system;

/// `workspace:restore(defaultConfig)`: return the saved workspace with its files
/// re-read from disk, or `default` when nothing was saved.
pub fn restore(config: &Config, default: Value) -> Value {
    let workspace = config
        .get("workspace", &[Value::Null])
        .unwrap_or(Value::Null);

    let Value::Object(mut workspace) = workspace else {
        return default;
    };

    let saved_files = match workspace.get("files") {
        Some(Value::Array(files)) => files.clone(),
        _ => Vec::new(),
    };

    let mut files = Vec::new();

    for file in &saved_files {
        let Some(path) = file.get("path").and_then(Value::as_str) else {
            continue;
        };

        // skip unreadable files (deleted/moved since save), like the Electron
        // try/catch around readFile
        if let Ok(descriptor) = file_system::read_file(path, &Value::Null) {
            files.push(descriptor);
        }
    }

    workspace.insert("files".to_string(), Value::Array(files));

    Value::Object(workspace)
}

/// `workspace:save(workspace)`: persist the workspace, returning `Null`
/// (`done(null)`).
pub fn save(config: &Config, workspace: Value) -> Result<Value, IpcError> {
    config.set("workspace", workspace)?;

    Ok(Value::Null)
}
